CI: Integration Tests - TestGCEnableRatchet

## CI failure

Link: https://github.com/cilium/cilium/actions/runs/18154090911/job/51670131670

```
--- FAIL: TestGCEnableRatchet (10.01s)

2025/10/01 07:08:46 INFO Starting GC of connection tracking first=true
2025/10/01 07:08:46 INFO Conntrack garbage collector interval recalculated expectedPrevInterval=5m0s actualPrevInterval=0s newInterval=7m30s deleteRatio=1e-17 adjustedDeleteRatio=1e-17
2025/10/01 07:08:46 INFO initial gc of ct and nat maps completed duration=485.302µs
2025/10/01 07:08:46 INFO Starting GC of connection tracking first=false
2025/10/01 07:08:47 INFO Starting GC of connection tracking first=false
2025/10/01 07:08:47 INFO Starting GC of connection tracking first=false
2025/10/01 07:08:48 INFO Starting GC of connection tracking first=false
2025/10/01 07:08:48 INFO Starting GC of connection tracking first=false
2025/10/01 07:08:49 INFO Starting GC of connection tracking first=false
2025/10/01 07:08:49 INFO Starting GC of connection tracking first=false
2025/10/01 07:08:50 INFO Starting GC of connection tracking first=false
2025/10/01 07:08:50 INFO Starting GC of connection tracking first=false
2025/10/01 07:08:51 INFO Starting GC of connection tracking first=false
    gc_test.go:183: 
        	Error Trace:	/home/runner/work/cilium/cilium/pkg/maps/ctmap/gc/gc_test.go:184
        	            				/opt/hostedtoolcache/go/1.25.1/x64/src/runtime/asm_amd64.s:1693
        	Error:      	"0" is not greater than "1"
    gc_test.go:183: 
        	Error Trace:	/home/runner/work/cilium/cilium/pkg/maps/ctmap/gc/gc_test.go:183
        	Error:      	Condition never satisfied
        	Test:       	TestGCEnableRatchet
2025/10/01 07:08:51 INFO Starting GC of connection tracking first=false
2025/10/01 07:08:52 INFO Starting GC of connection tracking first=false
2025/10/01 07:08:52 INFO Starting GC of connection tracking first=false
2025/10/01 07:08:53 INFO Starting GC of connection tracking first=false
2025/10/01 07:08:53 INFO Starting GC of connection tracking first=false
2025/10/01 07:08:54 INFO Starting GC of connection tracking first=false
2025/10/01 07:08:54 INFO Starting GC of connection tracking first=false
2025/10/01 07:08:55 INFO Starting GC of connection tracking first=false
2025/10/01 07:08:55 INFO Starting GC of connection tracking first=false
    gc_test.go:183: 
        	Error Trace:	/home/runner/work/cilium/cilium/pkg/maps/ctmap/gc/gc_test.go:184
        	            				/opt/hostedtoolcache/go/1.25.1/x64/src/runtime/asm_amd64.s:1693
        	Error:      	"0" is not greater than "0"
    gc_test.go:183: 
        	Error Trace:	/home/runner/work/cilium/cilium/pkg/maps/ctmap/gc/gc_test.go:183
        	Error:      	Condition never satisfied
        	Test:       	TestGCEnableRatchet
    gc_test.go:188: 
        	Error Trace:	/home/runner/work/cilium/cilium/pkg/maps/ctmap/gc/gc_test.go:188
        	Error:      	"5.000636943s" is not greater than "5.000687927s"
        	Test:       	TestGCEnableRatchet
2025/10/01 07:08:56 INFO Starting GC of connection tracking first=false
2025/10/01 07:08:56 INFO Conntrack garbage collector interval recalculated expectedPrevInterval=7m30s actualPrevInterval=10.001703933s newInterval=45s deleteRatio=0.9 adjustedDeleteRatio=40.49310024702168
2025/10/01 07:08:56 INFO Starting GC of connection tracking first=false
```

Easily reproduces locally as well, especially with race detection enabled (which reports a bunch of races):

```
2025/10/01 09:44:52 INFO Starting GC of connection tracking first=true
==================
WARNING: DATA RACE
Write at 0x00c0006c8c08 by goroutine 32:
  github.com/cilium/cilium/pkg/maps/ctmap/gc.TestGCEnableRatchet.func2()
      /home/marco/Documents/isovalent/cilium/pkg/maps/ctmap/gc/gc_test.go:164 +0xc4
  github.com/cilium/cilium/pkg/maps/ctmap/gc.(*GC).enable.func1()
      /home/marco/Documents/isovalent/cilium/pkg/maps/ctmap/gc/gc.go:198 +0x56e

Previous read at 0x00c0006c8c08 by goroutine 31:
  github.com/cilium/cilium/pkg/maps/ctmap/gc.TestGCEnableRatchet()
      /home/marco/Documents/isovalent/cilium/pkg/maps/ctmap/gc/gc_test.go:169 +0x5c9
  testing.tRunner()
      /home/marco/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.25.0.linux-amd64/src/testing/testing.go:1934 +0x21c
  testing.(*T).Run.gowrap1()
      /home/marco/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.25.0.linux-amd64/src/testing/testing.go:1997 +0x44

Goroutine 32 (running) created at:
  github.com/cilium/cilium/pkg/maps/ctmap/gc.(*GC).enable()
      /home/marco/Documents/isovalent/cilium/pkg/maps/ctmap/gc/gc.go:135 +0x17c
  github.com/cilium/cilium/pkg/maps/ctmap/gc.TestGCEnableRatchet()
      /home/marco/Documents/isovalent/cilium/pkg/maps/ctmap/gc/gc_test.go:154 +0x5bc
  testing.tRunner()
      /home/marco/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.25.0.linux-amd64/src/testing/testing.go:1934 +0x21c
  testing.(*T).Run.gowrap1()
      /home/marco/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.25.0.linux-amd64/src/testing/testing.go:1997 +0x44

Goroutine 31 (running) created at:
  testing.(*T).Run()
      /home/marco/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.25.0.linux-amd64/src/testing/testing.go:1997 +0x9d2
  testing.runTests.func1()
      /home/marco/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.25.0.linux-amd64/src/testing/testing.go:2477 +0x85
  testing.tRunner()
      /home/marco/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.25.0.linux-amd64/src/testing/testing.go:1934 +0x21c
  testing.runTests()
      /home/marco/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.25.0.linux-amd64/src/testing/testing.go:2475 +0x96c
  testing.(*M).Run()
      /home/marco/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.25.0.linux-amd64/src/testing/testing.go:2337 +0xed4
  main.main()
      _testmain.go:47 +0x164
==================
==================
WARNING: DATA RACE
Read at 0x000007088260 by goroutine 32:
  github.com/cilium/cilium/pkg/maps/ctmap.GetInterval()
      /home/marco/Documents/isovalent/cilium/pkg/maps/ctmap/ctmap.go:741 +0x131
  github.com/cilium/cilium/pkg/maps/ctmap/gc.(*GC).enable.func1()
      /home/marco/Documents/isovalent/cilium/pkg/maps/ctmap/gc/gc.go:201 +0x5ad

Previous write at 0x000007088260 by goroutine 31:
  github.com/cilium/cilium/pkg/maps/ctmap/gc.TestGCEnableRatchet()
      /home/marco/Documents/isovalent/cilium/pkg/maps/ctmap/gc/gc_test.go:170 +0x604
  testing.tRunner()
      /home/marco/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.25.0.linux-amd64/src/testing/testing.go:1934 +0x21c
  testing.(*T).Run.gowrap1()
      /home/marco/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.25.0.linux-amd64/src/testing/testing.go:1997 +0x44

Goroutine 32 (running) created at:
  github.com/cilium/cilium/pkg/maps/ctmap/gc.(*GC).enable()
      /home/marco/Documents/isovalent/cilium/pkg/maps/ctmap/gc/gc.go:135 +0x17c
  github.com/cilium/cilium/pkg/maps/ctmap/gc.TestGCEnableRatchet()
      /home/marco/Documents/isovalent/cilium/pkg/maps/ctmap/gc/gc_test.go:154 +0x5bc
  testing.tRunner()
      /home/marco/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.25.0.linux-amd64/src/testing/testing.go:1934 +0x21c
  testing.(*T).Run.gowrap1()
      /home/marco/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.25.0.linux-amd64/src/testing/testing.go:1997 +0x44

Goroutine 31 (running) created at:
  testing.(*T).Run()
      /home/marco/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.25.0.linux-amd64/src/testing/testing.go:1997 +0x9d2
  testing.runTests.func1()
      /home/marco/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.25.0.linux-amd64/src/testing/testing.go:2477 +0x85
  testing.tRunner()
      /home/marco/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.25.0.linux-amd64/src/testing/testing.go:1934 +0x21c
  testing.runTests()
      /home/marco/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.25.0.linux-amd64/src/testing/testing.go:2475 +0x96c
  testing.(*M).Run()
      /home/marco/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.25.0.linux-amd64/src/testing/testing.go:2337 +0xed4
  main.main()
      _testmain.go:47 +0x164
==================
2025/10/01 09:44:52 INFO Conntrack garbage collector interval recalculated expectedPrevInterval=5m0s actualPrevInterval=0s newInterval=7m30s deleteRatio=1e-17 adjustedDeleteRatio=1e-17
2025/10/01 09:44:52 INFO initial gc of ct and nat maps completed duration=1.480844ms
==================
WARNING: DATA RACE
Write at 0x00c00054a8ec by goroutine 31:
  github.com/cilium/cilium/pkg/maps/ctmap/gc.TestGCEnableRatchet()
      /home/marco/Documents/isovalent/cilium/pkg/maps/ctmap/gc/gc_test.go:186 +0x864
  testing.tRunner()
      /home/marco/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.25.0.linux-amd64/src/testing/testing.go:1934 +0x21c
  testing.(*T).Run.gowrap1()
      /home/marco/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.25.0.linux-amd64/src/testing/testing.go:1997 +0x44

Previous read at 0x00c00054a8ec by goroutine 10450:
  github.com/cilium/cilium/pkg/maps/ctmap/gc.TestGCEnableRatchet.func4()
      /home/marco/Documents/isovalent/cilium/pkg/maps/ctmap/gc/gc_test.go:184 +0x53
  github.com/stretchr/testify/assert.EventuallyWithT.func1()
      /home/marco/Documents/isovalent/cilium/vendor/github.com/stretchr/testify/assert/assertions.go:2096 +0xb3

Goroutine 31 (running) created at:
  testing.(*T).Run()
      /home/marco/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.25.0.linux-amd64/src/testing/testing.go:1997 +0x9d2
  testing.runTests.func1()
      /home/marco/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.25.0.linux-amd64/src/testing/testing.go:2477 +0x85
  testing.tRunner()
      /home/marco/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.25.0.linux-amd64/src/testing/testing.go:1934 +0x21c
  testing.runTests()
      /home/marco/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.25.0.linux-amd64/src/testing/testing.go:2475 +0x96c
  testing.(*M).Run()
      /home/marco/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.25.0.linux-amd64/src/testing/testing.go:2337 +0xed4
  main.main()
      _testmain.go:47 +0x164

Goroutine 10450 (finished) created at:
  github.com/stretchr/testify/assert.EventuallyWithT()
      /home/marco/Documents/isovalent/cilium/vendor/github.com/stretchr/testify/assert/assertions.go:2119 +0x3a7
  github.com/cilium/cilium/pkg/maps/ctmap/gc.TestGCEnableRatchet()
      /home/marco/Documents/isovalent/cilium/pkg/maps/ctmap/gc/gc_test.go:183 +0x844
  testing.tRunner()
      /home/marco/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.25.0.linux-amd64/src/testing/testing.go:1934 +0x21c
  testing.(*T).Run.gowrap1()
      /home/marco/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.25.0.linux-amd64/src/testing/testing.go:1997 +0x44
==================
2025/10/01 09:45:12 INFO Starting GC of connection tracking first=false
==================
WARNING: DATA RACE
Read at 0x00c0004aaf90 by goroutine 32:
  github.com/cilium/cilium/pkg/maps/ctmap/gc.TestGCEnableRatchet.func2()
      /home/marco/Documents/isovalent/cilium/pkg/maps/ctmap/gc/gc_test.go:166 +0xf8
  github.com/cilium/cilium/pkg/maps/ctmap/gc.(*GC).enable.func1()
      /home/marco/Documents/isovalent/cilium/pkg/maps/ctmap/gc/gc.go:198 +0x56e

Previous write at 0x00c0004aaf90 by goroutine 31:
  github.com/cilium/cilium/pkg/maps/ctmap/gc.TestGCEnableRatchet()
      /home/marco/Documents/isovalent/cilium/pkg/maps/ctmap/gc/gc_test.go:200 +0x967
  testing.tRunner()
      /home/marco/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.25.0.linux-amd64/src/testing/testing.go:1934 +0x21c
  testing.(*T).Run.gowrap1()
      /home/marco/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.25.0.linux-amd64/src/testing/testing.go:1997 +0x44

Goroutine 32 (running) created at:
  github.com/cilium/cilium/pkg/maps/ctmap/gc.(*GC).enable()
      /home/marco/Documents/isovalent/cilium/pkg/maps/ctmap/gc/gc.go:135 +0x17c
  github.com/cilium/cilium/pkg/maps/ctmap/gc.TestGCEnableRatchet()
      /home/marco/Documents/isovalent/cilium/pkg/maps/ctmap/gc/gc_test.go:154 +0x5bc
  testing.tRunner()
      /home/marco/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.25.0.linux-amd64/src/testing/testing.go:1934 +0x21c
  testing.(*T).Run.gowrap1()
      /home/marco/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.25.0.linux-amd64/src/testing/testing.go:1997 +0x44

Goroutine 31 (running) created at:
  testing.(*T).Run()
      /home/marco/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.25.0.linux-amd64/src/testing/testing.go:1997 +0x9d2
  testing.runTests.func1()
      /home/marco/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.25.0.linux-amd64/src/testing/testing.go:2477 +0x85
  testing.tRunner()
      /home/marco/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.25.0.linux-amd64/src/testing/testing.go:1934 +0x21c
  testing.runTests()
      /home/marco/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.25.0.linux-amd64/src/testing/testing.go:2475 +0x96c
  testing.(*M).Run()
      /home/marco/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.25.0.linux-amd64/src/testing/testing.go:2337 +0xed4
  main.main()
      _testmain.go:47 +0x164
==================
2025/10/01 09:45:12 INFO Conntrack garbage collector interval recalculated expectedPrevInterval=7m30s actualPrevInterval=20.005266533s newInterval=45s deleteRatio=0.9 adjustedDeleteRatio=20.24466903912157
--- FAIL: TestGCEnableRatchet (20.02s)
    gc_test.go:172: 
        	Error Trace:	/home/marco/Documents/isovalent/cilium/pkg/maps/ctmap/gc/gc_test.go:173
        	            				/home/marco/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.25.0.linux-amd64/src/runtime/asm_amd64.s:1693
        	Error:      	Should not be zero, but was 0
    gc_test.go:172: 
        	Error Trace:	/home/marco/Documents/isovalent/cilium/pkg/maps/ctmap/gc/gc_test.go:172
        	Error:      	Condition never satisfied
        	Test:       	TestGCEnableRatchet
    gc_test.go:183: 
        	Error Trace:	/home/marco/Documents/isovalent/cilium/pkg/maps/ctmap/gc/gc_test.go:184
        	            				/home/marco/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.25.0.linux-amd64/src/runtime/asm_amd64.s:1693
        	Error:      	"0" is not greater than "0"
    gc_test.go:183: 
        	Error Trace:	/home/marco/Documents/isovalent/cilium/pkg/maps/ctmap/gc/gc_test.go:183
        	Error:      	Condition never satisfied
        	Test:       	TestGCEnableRatchet
    gc_test.go:183: 
        	Error Trace:	/home/marco/Documents/isovalent/cilium/pkg/maps/ctmap/gc/gc_test.go:184
        	            				/home/marco/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.25.0.linux-amd64/src/runtime/asm_amd64.s:1693
        	Error:      	"0" is not greater than "0"
    gc_test.go:183: 
        	Error Trace:	/home/marco/Documents/isovalent/cilium/pkg/maps/ctmap/gc/gc_test.go:183
        	Error:      	Condition never satisfied
        	Test:       	TestGCEnableRatchet
    gc_test.go:188: 
        	Error Trace:	/home/marco/Documents/isovalent/cilium/pkg/maps/ctmap/gc/gc_test.go:188
        	Error:      	"5.001433313s" is not greater than "5.001553401s"
        	Test:       	TestGCEnableRatchet
    testing.go:1617: race detected during execution of test
FAIL
```

/cc @tommyp1ckles

--- a/pkg/maps/ctmap/ctmap.go
+++ b/pkg/maps/ctmap/ctmap.go
@@ -726,7 +726,13 @@ func Exists(ipv4, ipv6 bool) bool {
 //   - expectedPrevInterval 	= Is the last computed interval, which we expected to
 //     wait *unless* a signal caused early pass. If this is set to zero then we use gc starting interval.
 func GetInterval(logger *slog.Logger, actualPrevInterval, expectedPrevInterval time.Duration, maxDeleteRatio float64) time.Duration {
-	if val := option.Config.ConntrackGCInterval; val != time.Duration(0) {
+	return GetIntervalWithConfig(logger, actualPrevInterval, expectedPrevInterval, maxDeleteRatio,
+		option.Config.ConntrackGCInterval, option.Config.ConntrackGCMaxInterval, GCIntervalRounding, MinGCInterval)
+}
+
+func GetIntervalWithConfig(logger *slog.Logger, actualPrevInterval, expectedPrevInterval time.Duration, maxDeleteRatio float64,
+	conntrackGCInterval, conntrackGCMaxInterval, gcIntervalRounding, minGCInterval time.Duration) time.Duration {
+	if val := conntrackGCInterval; val != time.Duration(0) {
 		return val
 	}
 
@@ -737,8 +743,8 @@ func GetInterval(logger *slog.Logger, actualPrevInterval, expectedPrevInterval t
 		adjustedDeleteRatio *= float64(expectedPrevInterval) / float64(actualPrevInterval)
 	}
 
-	newInterval := calculateInterval(expectedPrevInterval, adjustedDeleteRatio)
-	if val := option.Config.ConntrackGCMaxInterval; val != time.Duration(0) && newInterval > val {
+	newInterval := calculateIntervalWithConfig(expectedPrevInterval, adjustedDeleteRatio, gcIntervalRounding, minGCInterval)
+	if val := conntrackGCMaxInterval; val != time.Duration(0) && newInterval > val {
 		newInterval = val
 	}
 
@@ -764,6 +770,10 @@ var (
 )
 
 func calculateInterval(prevInterval time.Duration, maxDeleteRatio float64) (interval time.Duration) {
+	return calculateIntervalWithConfig(prevInterval, maxDeleteRatio, GCIntervalRounding, MinGCInterval)
+}
+
+func calculateIntervalWithConfig(prevInterval time.Duration, maxDeleteRatio float64, gcIntervalRounding, minGCInterval time.Duration) (interval time.Duration) {
 	interval = prevInterval
 
 	if maxDeleteRatio == 0.0 {
@@ -776,13 +786,13 @@ func calculateInterval(prevInterval time.Duration, maxDeleteRatio float64) (inte
 			maxDeleteRatio = 0.9
 		}
 		// 25%..90% => 1.3x..10x shorter
-		interval = max(time.Duration(float64(interval)*(1.0-maxDeleteRatio)).Round(GCIntervalRounding), MinGCInterval)
+		interval = max(time.Duration(float64(interval)*(1.0-maxDeleteRatio)).Round(gcIntervalRounding), minGCInterval)
 	case maxDeleteRatio < 0.05:
 		// When less than 5% of entries were deleted, increase the
 		// interval. Use a simple 1.5x multiplier to start growing slowly
 		// as a new node may not be seeing workloads yet and thus the
 		// scan will return a low deletion ratio at first.
-		interval = min(time.Duration(float64(interval)*1.5).Round(GCIntervalRounding), defaults.ConntrackGCMaxLRUInterval)
+		interval = min(time.Duration(float64(interval)*1.5).Round(gcIntervalRounding), defaults.ConntrackGCMaxLRUInterval)
 	}
 
 	return

--- a/pkg/maps/ctmap/gc/gc.go
+++ b/pkg/maps/ctmap/gc/gc.go
@@ -157,6 +157,16 @@ func (gc *GC) enable(
 	runGC func(ipv4, ipv6, triggeredBySignal bool, filter ctmap.GCFilter) (maxDeleteRatio float64, success bool),
 	runMapPressureDaemon bool,
 ) {
+	gc.enableWithConfig(runGC, runMapPressureDaemon,
+		option.Config.ConntrackGCInterval, option.Config.ConntrackGCMaxInterval,
+		ctmap.GCIntervalRounding, ctmap.MinGCInterval)
+}
+
+func (gc *GC) enableWithConfig(
+	runGC func(ipv4, ipv6, triggeredBySignal bool, filter ctmap.GCFilter) (maxDeleteRatio float64, success bool),
+	runMapPressureDaemon bool,
+	conntrackGCInterval, conntrackGCMaxInterval, gcIntervalRounding, minGCInterval time.Duration,
+) {
 	var (
 		initialScan         = true
 		initialScanComplete = make(chan struct{})
@@ -229,7 +239,8 @@ func (gc *GC) enable(
 				maxDeleteRatio, success = runGC(ipv4, ipv6, triggeredBySignal, gcFilter)
 			}
 
-			interval := ctmap.GetInterval(gc.logger, gcInterval, cachedGCInterval, maxDeleteRatio)
+			interval := ctmap.GetIntervalWithConfig(gc.logger, gcInterval, cachedGCInterval, maxDeleteRatio,
+				conntrackGCInterval, conntrackGCMaxInterval, gcIntervalRounding, minGCInterval)
 			if success && gc.isFullGC(ipv4, ipv6) {
 				// Mark the CT GC as over in each EP DNSZombies instance, if we did a *full* GC run
 				nextGCTime := time.Now().Add(interval)

--- a/pkg/maps/ctmap/gc/gc_test.go
+++ b/pkg/maps/ctmap/gc/gc_test.go
@@ -6,6 +6,7 @@ package gc
 import (
 	"context"
 	"log/slog"
+	"runtime"
 	"sync/atomic"
 	"testing"
 	"time"
@@ -15,20 +16,22 @@ import (
 	"github.com/cilium/cilium/pkg/endpoint"
 	"github.com/cilium/cilium/pkg/fqdn"
 	"github.com/cilium/cilium/pkg/maps/ctmap"
-	"github.com/cilium/cilium/pkg/option"
 	"github.com/cilium/cilium/pkg/signal"
 )
 
 // TestGCEnable tests the overall *flow* of ctmap GC scheduling
 // while abstracting out the underlying garbage collection logic.
 func TestGCEnableDualStack(t *testing.T) {
+	signalChan := make(chan SignalData)
+	defer close(signalChan) // Ensure cleanup
+
 	gc := &GC{
 		ipv4:             true,
 		ipv6:             true,
 		logger:           slog.Default(),
 		endpointsManager: &fakeEPM{},
 		signalHandler: SignalHandler{
-			signals: make(chan SignalData),
+			signals: signalChan,
 			manager: &fakeSignalMan{},
 		},
 	}
@@ -44,8 +47,12 @@ func TestGCEnableDualStack(t *testing.T) {
 	}
 
 	returnRatio := 0.1 // low -> high next interval
-	option.Config.ConntrackGCMaxInterval = time.Millisecond * 500
-	gc.enable(func(ipv4, ipv6, triggeredBySignal bool, filter ctmap.GCFilter) (maxDeleteRatio float64, success bool) {
+	// Use local variables instead of modifying global state
+	localConntrackGCMaxInterval := time.Millisecond * 500
+	localGCIntervalRounding := ctmap.GCIntervalRounding
+	localMinGCInterval := ctmap.MinGCInterval
+
+	gc.enableWithConfig(func(ipv4, ipv6, triggeredBySignal bool, filter ctmap.GCFilter) (maxDeleteRatio float64, success bool) {
 		if ipv4 {
 			ipv4Passes.Add(1)
 			if ipv6 {
@@ -53,14 +60,14 @@ func TestGCEnableDualStack(t *testing.T) {
 			}
 		}
 		return returnRatio, true
-	}, false)
+	}, false, 0, localConntrackGCMaxInterval, localGCIntervalRounding, localMinGCInterval)
 
 	assert.EventuallyWithT(t, func(c *assert.CollectT) {
 		assert.Equal(c, 1, int(dualPasses.Load()))
 		assert.Equal(c, 1, int(ipv4Passes.Load()))
 	}, time.Second, time.Millisecond*10, "initial pass should be full pass")
 
-	gc.signalHandler.signals <- SignalProtoV4
+	signalChan <- SignalProtoV4
 
 	assert.EventuallyWithT(t, func(c *assert.CollectT) {
 		assert.Equal(c, 1, int(dualPasses.Load()))
@@ -119,94 +126,131 @@ func TestGCEnableDualStack(t *testing.T) {
 }
 
 // TestGCEnableRatchet tests the behavior of low, then high, purge ratios
-// causing the interval to ratched up/down in successive GC passes.
+// causing the interval to ratchet up/down in successive GC passes.
 func TestGCEnableRatchet(t *testing.T) {
-	gc := &GC{
-		ipv4:             true,
-		ipv6:             true,
-		logger:           slog.Default(),
-		endpointsManager: &fakeEPM{},
-		signalHandler: SignalHandler{
-			signals: make(chan SignalData),
-			manager: &fakeSignalMan{},
-		},
-	}
+	// Use local configuration variables instead of modifying global state
+	localGCIntervalRounding := 10 * time.Millisecond
+	localMinGCInterval := time.Millisecond
+	localConntrackGCMaxInterval := 2 * time.Second // Set reasonable max for testing
 
-	// start GC
-	var ipv4Passes atomic.Int32
-	var dualPasses atomic.Int32
+	// Channel to signal test completion to ensure proper cleanup ordering
+	testDone := make(chan struct{})
 
-	reset := func() {
-		ipv4Passes.Store(0)
-		dualPasses.Store(0)
-	}
+	defer func() {
+		close(testDone)
+		// Wait to ensure goroutines fully exit
+		time.Sleep(200 * time.Millisecond)
+		// Force garbage collection and yield to help ensure goroutines are done
+		runtime.GC()
+		runtime.Gosched()
+		time.Sleep(50 * time.Millisecond)
+	}()
 
-	ctmap.GCIntervalRounding = 10 * time.Millisecond
-	returnRatio := 0.00000000000000001 // low -> high next interval
-	option.Config.ConntrackGCMaxInterval = 500 * time.Millisecond
-	initialPass := make(chan struct{})
-	// 1. First we allow at least one initial pass to happen with the following conditions:
-	// * ConntrackGCMaxInterval = Seconod
-	// * Delete Ratio -> very small (meaning we quickly increase next interval)
-	//	Note: This does not affect the first interval, as that runs right away
-	//	instead a first pass must be run that returns this value.
-	// 	The next interval then uses this to compute next interval duration.
-	gc.enable(func(ipv4, ipv6, triggeredBySignal bool, filter ctmap.GCFilter) (maxDeleteRatio float64, success bool) {
-		if ipv4 {
-			ipv4Passes.Add(1)
-			if ipv6 {
-				dualPasses.Add(1)
+	// Test interval ratcheting down then up
+	{
+		signalChan := make(chan SignalData, 10)
+		gc := &GC{
+			ipv4:             true,
+			ipv6:             true,
+			logger:           slog.Default(),
+			endpointsManager: &fakeEPM{},
+			signalHandler: SignalHandler{
+				signals: signalChan,
+				manager: &fakeSignalMan{},
+			},
+		}
+
+		var passCount atomic.Int32
+		var lastPassTime time.Time
+		var lastRealDuration atomic.Int64 // Store as nanoseconds
+		var currentDeleteRatio atomic.Pointer[float64]
+
+		initialRatio := 0.9
+		currentDeleteRatio.Store(&initialRatio)
+
+		initialPassDone := make(chan struct{})
+		var initialPassSignaled atomic.Bool
+
+		gc.enableWithConfig(func(ipv4, ipv6, triggeredBySignal bool, filter ctmap.GCFilter) (maxDeleteRatio float64, success bool) {
+			// Check if test is done
+			select {
+			case <-testDone:
+				return 0, false
+			default:
 			}
-		}
-		if initialPass != nil {
-			close(initialPass)
-			reset()
-			initialPass = nil
-		}
-		return returnRatio, true
-	}, false)
 
-	<-initialPass                            // wait for initial pass to complete.
-	option.Config.ConntrackGCMaxInterval = 0 // allow for large interval but expect shortk
+			if ipv4 {
+				passCount.Add(1)
+				now := time.Now()
+				if !lastPassTime.IsZero() {
+					duration := now.Sub(lastPassTime)
+					lastRealDuration.Store(int64(duration))
+				}
+				lastPassTime = now
+			}
 
-	assert.EventuallyWithT(t, func(c *assert.CollectT) {
-		assert.NotZero(c, int(ipv4Passes.Load()))
-	}, 10*time.Second, time.Millisecond)
+			// Signal initial pass completion once
+			if !initialPassSignaled.Swap(true) {
+				close(initialPassDone)
+			}
 
-	prev := ipv4Passes.Load()
-	lastPassTime := time.Now()
-	var lastRealDuration time.Duration
-	// Wait for the gc to happen 3 times, each time we expect the interval
-	// to ratched up (note: these are rounded up to seconds, so to avoid
-	// making tests run too long we just wait for three iterations).
-	for range 2 {
+			ratio := *currentDeleteRatio.Load()
+			return ratio, true
+		}, false, 0, localConntrackGCMaxInterval, localGCIntervalRounding, localMinGCInterval)
+
+		// Wait for initial pass
+		<-initialPassDone
+
+		// Phase 1: Ratchet DOWN (High delete ratio 0.9)
+		t.Log("Phase 1: Ratcheting DOWN")
+
+		var prevDuration time.Duration
+		initialCount := passCount.Load()
 		assert.EventuallyWithT(t, func(c *assert.CollectT) {
-			assert.Greater(c, ipv4Passes.Load(), prev)
-		}, time.Second*5, time.Millisecond*10)
-		prev = ipv4Passes.Load()
-		if lastRealDuration != 0 {
-			assert.Greater(t, time.Since(lastPassTime), lastRealDuration)
-		}
-		lastRealDuration = time.Since(lastPassTime)
-		lastPassTime = time.Now()
+			assert.Greater(c, passCount.Load(), initialCount)
+
+			currentDuration := time.Duration(lastRealDuration.Load())
+			if currentDuration > 0 {
+				if prevDuration == 0 {
+					prevDuration = currentDuration
+				} else if currentDuration < prevDuration {
+					prevDuration = currentDuration
+				}
+				// Verify it drops close to min interval
+				assert.LessOrEqual(c, currentDuration, localMinGCInterval*200, "Interval should drop significantly") // 200ms
+			}
+		}, 5*time.Second, 10*time.Millisecond)
+
+		// Phase 2: Ratchet UP (Low delete ratio 0.01)
+		t.Log("Phase 2: Ratcheting UP")
+		lowRatio := 0.01
+		currentDeleteRatio.Store(&lowRatio)
+
+		// Reset prevDuration to track increase from the low point
+		prevDuration = 0
+		initialCount = passCount.Load()
+
+		assert.EventuallyWithT(t, func(c *assert.CollectT) {
+			assert.Greater(c, passCount.Load(), initialCount)
+
+			currentDuration := time.Duration(lastRealDuration.Load())
+			if currentDuration > 0 {
+				if prevDuration == 0 {
+					prevDuration = currentDuration
+				} else if currentDuration > prevDuration {
+					prevDuration = currentDuration
+				}
+
+				// Verify it increases.
+				assert.Greater(c, currentDuration, localMinGCInterval*200, "Interval should increase back up")
+			}
+		}, 5*time.Second, 10*time.Millisecond)
+
+		// Stop
+		close(signalChan)
 	}
 
-	reset()
-	ctmap.MinGCInterval = time.Millisecond
-	gc.signalHandler.signals <- SignalProtoV4
-	// make purge ratio very big -> ratched down intervals.
-	// This ratchets *fast* and will go to the min value (i.e. MinGCInterval)
-	// very fast, so we just do one iteration.
-	returnRatio = 0.9
-	lastRealDuration = 0
-	lastPassTime = time.Now()
-	prev = ipv4Passes.Load()
-
-	assert.EventuallyWithT(t, func(c *assert.CollectT) {
-		assert.Greater(c, ipv4Passes.Load(), prev)
-	}, time.Second*5, time.Millisecond*10)
-	prev = ipv4Passes.Load()
-	assert.Greater(t, time.Since(lastPassTime), lastRealDuration)
+	t.Log("GC interval ratcheting test completed successfully")
 }
 
 type fakeEPM struct{}

